include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -918,6 +995,587 @@ namespace internal
+            }
+
+          // grad y
+          eval0.template values<0, true, false, normal_dir>(values_dofs, temp1);
+          if (evaluation_flag & EvaluationFlags::gradients)
+            {
+              eval1.template gradients<1, true, false, normal_dir>(temp1,
+                                                                   temp2);
+              eval2.template values<2, true, false, normal_dir>(temp2,
+                                                                gradients_quad +
+                                                                  n_q_points);
+            }
+
+          if (evaluation_flag & EvaluationFlags::hessians)
+            {
+              // grad yz
+              if (!(evaluation_flag & EvaluationFlags::gradients))
+                eval1.template gradients<1, true, false, normal_dir>(temp1,
+                                                                     temp2);
+              eval2.template gradients<2, true, false, normal_dir>(
+                temp2, hessians_quad + 5 * n_q_points);
+
+              // grad yy
+              eval1.template hessians<1, true, false, normal_dir>(temp1, temp2);
+              eval2.template values<2, true, false, normal_dir>(temp2,
+                                                                hessians_quad +
+                                                                  n_q_points);
+            }
+
+          // grad z: can use the values applied in x direction stored in
+          // temp1
+          eval1.template values<1, true, false, normal_dir>(temp1, temp2);
+          if (evaluation_flag & EvaluationFlags::gradients)
+            eval2.template gradients<2, true, false, normal_dir>(
+              temp2, gradients_quad + 2 * n_q_points);
+
+          // grad zz: can use the values applied in x and y direction stored
+          // in temp2
+          if (evaluation_flag & EvaluationFlags::hessians)
+            eval2.template hessians<2, true, false, normal_dir>(
+              temp2, hessians_quad + 2 * n_q_points);
+
+          // val: can use the values applied in x & y direction stored in
+          // temp2
+          if (evaluation_flag & EvaluationFlags::values)
+            eval2.template values<2, true, false, normal_dir>(temp2,
+                                                              values_quad);
+          break;
+        default:
+          AssertThrow(false, ExcNotImplemented());
+      }
+  }
+
+  template <int dim, int fe_degree, int n_q_points_1d, typename Number>
+  inline void
+  FEEvaluationImpl<
+    MatrixFreeFunctions::tensor_raviart_thomas,
+    dim,
+    fe_degree,
+    n_q_points_1d,
+    Number>::integrate(const EvaluationFlags::EvaluationFlags integration_flag,
+                       Number *                              values_dofs_actual,
+                       FEEvaluationData<dim, Number, false> &fe_eval,
+                       const bool add_into_values_array)
```

Comment:
Looking at the implementation of this function, this feels very repetitive from `evaluate`. I wonder if there is a way to unify them into the same function and branch into evaluate/integrate by a suitable function. I don't think have a suggestion for a change now and would go with the present code, but if anyone has a good idea this would remove 50 lines of code.

Reply:
Yes the only difference would be the input parameter `add_into_values_array` to `integrate()`, where you could add a dummy variable to the `evaluate()` path, and then the calls to `evaluate/integrate_tensor_product_per_component()`. The distinction could be done by another template parameter.

Reply:
Realized after fighting with the compiler all day that `values_dofs_actual` is `const` for the `evaluate` path and I don't know how one could get around this.

Reply:
We can do a `const_cast` from `const Number*` to `Number *` if necessary (knowing that we do not modify things), or we can post-pone this step for now, as long as we make sure to not miss the low hanging fruits (or ideas expressed by @peterrum in another post). In my opinion it is not mandatory to fix this duplication problem at once, as there are some further steps in the pipeline.